### PR TITLE
[Kernel] Matrix Union

### PR DESCRIPTION
**Pull request by classicsong**
## Description
Make a union of two Matrix (coo or csr). The matrixes should have same shape (num_rows and num_cols).

e.g.,
```
/*
 * A = [[0, 0, 1, 0],
 *      [1, 0, 1, 1],
 *      [0, 1, 0, 0]]
 *
 * B = [[0, 1, 1, 0],
 *      [0, 0, 0, 1],
 *      [0, 0, 1, 0]]
 *
 * C = UnionCsr({A, B});
 * C = [[0, 1, 2, 0],
 *      [1, 0, 1, 2],
 *      [0, 1, 1, 0]]
 */
```

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

File: src/array/array.cc [hunk 1]
```diff
@@ -509,6 +501,32 @@ COOMatrix CSRRowWiseTopk(
...
+  CSRMatrix csr0 = (!CSRIsSorted(csrs[0])) ?
+    CSRSort(csrs[0]) :
+    csrs[0];
+  CSRMatrix csr1 = (!CSRIsSorted(csrs[1])) ?
+    CSRSort(csrs[1]) :
+    csrs[1];
+
+  const std::vector<CSRMatrix> input_csrs({csr0, csr1});
+  ATEN_CSR_SWITCH(csrs[0], XPU, IdType, "UnionCsr", {
+    ret = impl::UnionCsr<XPU, IdType>(input_csrs);
```
Comment (jermainewang) [inline src/array/array.cc:525]: You could just use `impl::UnionCsr<XPU, IdType>({csr0, csr1});`

File: src/array/cpu/csr_union.cc [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+namespace aten {
+namespace impl {
+
+template <DLDeviceType XPU, typename IdType>
+std::tuple<CSRMatrix, IdArray, IdArray>
+UnionCsr(const std::vector<CSRMatrix>& csrs) {
+  CHECK_EQ(csrs[0].num_rows, csrs[1].num_rows) <<
+    "UnionCsr requires CSRs to have the same shape";
+  CHECK_EQ(csrs[0].num_cols, csrs[1].num_cols) <<
+    "UnionCsr requires CSRs to have the same shape";
```
Comment (jermainewang) [inline src/array/cpu/csr_union.cc:24]: These checks can be omitted since they have been done before.

File: src/array/cpu/csr_union.cc [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+  CSRMatrix csr0 = csrs[0].sorted ? csrs[0] : CSRSort(csrs[0]);
+  CSRMatrix csr1 = csrs[1].sorted ? csrs[1] : CSRSort(csrs[1]);
```
Comment (jermainewang) [inline src/array/cpu/csr_union.cc:26]: Same here.

File: src/array/cpu/csr_union.cc [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+  std::vector<IdType> indptr;
+  std::vector<IdType> indices;
+  const int64_t nnz0 = csr0.indices->shape[0];
+  const int64_t nnz1 = csr1.indices->shape[0];
+  const IdType *indptr_data0 = static_cast<IdType*>(csr0.indptr->data);
+  const IdType *indptr_data1 = static_cast<IdType*>(csr1.indptr->data);
+  const IdType *indices_data0 = static_cast<IdType*>(csr0.indices->data);
+  const IdType *indices_data1 = static_cast<IdType*>(csr1.indices->data);
```
Comment (jermainewang) [inline src/array/cpu/csr_union.cc:35]: BTW, there is a new syntax for this: `csr0.Ptr<IdType>()`

**Comment by classicsong** [comment]
> Which user-facing API is this operator for?

We can use Reverse + Union (+ToSimple) to implement ToBidirected.

File: src/array/array.cc [hunk 5]
```diff
@@ -509,6 +501,32 @@ COOMatrix CSRRowWiseTopk(
   return ret;
 }
 
+std::tuple<CSRMatrix, IdArray, IdArray>
+UnionCsr(const std::vector<CSRMatrix>& csrs) {
+  std::tuple<CSRMatrix, IdArray, IdArray> ret;
+  CHECK_EQ(csrs.size(), 2) << "UnionCsr creates a union of two CSRMatrix";
```
[the remark this answers is not in the document]
Comment (classicsong) [reply src/array/array.cc:507]: Currently not. Any use case?

File: src/array/cpu/csr_union.cc [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+namespace impl {
+
+template <DLDeviceType XPU, typename IdType>
+CSRMatrix UnionCsr(const std::vector<CSRMatrix>& csrs) {
+  std::vector<IdType> res_indptr;
+  std::vector<IdType> res_indices;
+  std::vector<IdType> res_data;
+
+  // some preprocess
+  // we assume the size of csrs is not large in common cases
```
Comment (jermainewang) [inline src/array/cpu/csr_union.cc:24]: The number of csrs…

File: src/array/array.cc [hunk 7]
```diff
@@ -678,6 +697,66 @@ std::pair<COOMatrix, IdArray> COOCoalesce(COOMatrix coo) {
...
+  for (size_t i = 1; i < coos.size(); ++i) {
+    CHECK_EQ(coos[0].num_rows, coos[i].num_rows) <<
+      "UnionCoo requires both COOMatrix have same number of rows";
+    CHECK_EQ(coos[0].num_cols, coos[i].num_cols) <<
+      "UnionCoo requires both COOMatrix have same number of cols";
+    CHECK_SAME_CONTEXT(coos[0].row, coos[i].row);
+    CHECK_SAME_DTYPE(coos[0].row, coos[i].row);
+  }
+
+  // we assume the size of csrs is not large in common cases
```
Comment (jermainewang) [inline src/array/array.cc:713]: number of coos…

**Pull request merged**